### PR TITLE
Modify all_reduce to take `group`

### PR DESCRIPTION
**Pull request by puhuk**
Description: Modify all_reduce to take `group`

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/distributed/utils.py [hunk 1]
```diff
@@ -321,7 +321,9 @@ def train_fn(local_rank, a, b, c, d=12):
         )
 
 
-def all_reduce(tensor: Union[torch.Tensor, float], op: str = "SUM") -> Union[torch.Tensor, float]:
+def all_reduce(
+    tensor: Union[torch.Tensor, float], op: str = "SUM", group: Optional[Union[Any, List[int]]] = None
+) -> Union[torch.Tensor, float]:
```
Comment (vfdev-5) [inline ignite/distributed/utils.py:326]: We need update the docstring and add group arg

File: ignite/distributed/utils.py [hunk 2]
```diff
@@ -337,7 +340,10 @@ def all_reduce(tensor: Union[torch.Tensor, float], op: str = "SUM") -> Union[tor
     if _need_to_sync and isinstance(_model, _SerialModel):
         sync(temporary=True)
 
-    return _model.all_reduce(tensor, op)
+    if isinstance(group, list) and all(isinstance(item, int) for item in group):
```
Comment (puhuk) [inline ignite/distributed/utils.py:343]: loose the condition

File: ignite/distributed/comp_models/base.py [hunk 3]
```diff
@@ -269,7 +271,9 @@ def broadcast(
         return tensor
 
     @abstractmethod
-    def _do_all_reduce(self, tensor: torch.Tensor, op: str = "SUM") -> torch.Tensor:
+    def _do_all_reduce(
+        self, tensor: torch.Tensor, op: str = "SUM", group: Optional[Union[Any, List[int]]] = None
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/base.py:275]: Check `group` signature here as well, should it be `Optional[Union[Any, List[int]]]` ?

File: ignite/distributed/comp_models/base.py [hunk 4]
```diff
@@ -357,7 +363,9 @@ def broadcast(
             raise ValueError("Argument tensor should not be None")
         return tensor
 
-    def _do_all_reduce(self, tensor: torch.Tensor, op: str = "SUM") -> torch.Tensor:
+    def _do_all_reduce(
+        self, tensor: torch.Tensor, op: str = "SUM", group: Optional[Union[Any, List[int]]] = None
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/base.py:367]: Same here

File: ignite/distributed/comp_models/horovod.py [hunk 5]
```diff
@@ -165,13 +164,17 @@ def spawn(  # type: ignore[override]
...
             if op in self._manual_reduce_op_map:
                 op_fn = self._manual_reduce_op_map[op]
                 return self._do_manual_all_reduce(tensor, op_fn)
             if op not in self._reduce_op_map:
                 raise ValueError(f"Unsupported reduction operation: '{op}'")
             op = self._reduce_op_map[op]
+            if group is not None and not isinstance(group, hvd.ProcessSet):
+                raise ValueError("Group should be list of int or ProcessSet")
+            if group is not None:
+                return hvd.allreduce(tensor, op=op, process_set=group)
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/horovod.py:177]: Let's unify the code:

```suggestion
            if group is not None:
                if not isinstance(group, hvd.ProcessSet):
                    raise ValueError("Group should be list of int or ProcessSet")
                return hvd.allreduce(tensor, op=op, process_set=group)
```

File: ignite/distributed/comp_models/base.py [hunk 6]
```diff
@@ -342,7 +346,9 @@ def create_from_backend(backend: Optional[str] = None, **kwargs: Any) -> "_Seria
     def spawn(*args: Any, **kwargs: Any) -> None:
         raise NotImplementedError("Serial computation model does not implement spawn method")
 
-    def all_reduce(self, tensor: Union[torch.Tensor, float], op: str = "SUM") -> Union[torch.Tensor, float]:
+    def all_reduce(
+        self, tensor: Union[torch.Tensor, float], op: str = "SUM", group: Optional[Union[Any, List[int]]] = None
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/base.py:350]: Same here, check group type hint. Here is model all_reduce.

File: tests/ignite/distributed/utils/__init__.py [hunk 7]
```diff
@@ -119,6 +119,27 @@ def _test_distrib_all_reduce(device):
...
+        assert res == torch.tensor([sum(ranks)])
+
+        t = torch.tensor([rank], device=device)
+        res = idist.all_reduce(t, group=ranks)
+        assert res == torch.tensor([sum(ranks)])
+
+        ranks = "abc"
+        with pytest.raises(ValueError, match=r"Argument ranks should be list of int"):
+            group = idist.new_group(ranks)
+            res = idist.all_reduce(t, group=group)
```
Comment (vfdev-5) [inline tests/ignite/distributed/utils/__init__.py:140]: test cases should be: 
- group arg as ranks [0, 1]
- group arg as a backend group created with `idist.new_group`
- invalid group arg like "abc"

File: ignite/distributed/comp_models/horovod.py [hunk 8]
```diff
@@ -165,13 +164,18 @@ def spawn(  # type: ignore[override]
...
             if op in self._manual_reduce_op_map:
                 op_fn = self._manual_reduce_op_map[op]
                 return self._do_manual_all_reduce(tensor, op_fn)
             if op not in self._reduce_op_map:
                 raise ValueError(f"Unsupported reduction operation: '{op}'")
             op = self._reduce_op_map[op]
+            if group is not None:
+                if not isinstance(group, hvd.ProcessSet):
+                    raise ValueError("Argument group should be list of int or ProcessSet")
+                hvd.init(process_sets=[group])
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/horovod.py:177]: @puhuk do you know what `hvd.init(process_sets=[group])` does ?

Comment (vfdev-5) [reply ignite/distributed/comp_models/horovod.py:177]: and how it works if horovod is already initialized ?

Comment (puhuk) [reply ignite/distributed/comp_models/horovod.py:177]: It sets up the horovod backend again by calling `_setup_process_sets`

Comment (vfdev-5) [reply ignite/distributed/comp_models/horovod.py:177]: you have to check that

Comment (puhuk) [reply ignite/distributed/comp_models/horovod.py:177]: There are two options in initializing horovod
1. Static initialization like `hvd.init(process_sets= ~~)`   --> In this case, when initialized, process_set can not be changed
2. Dynamic process set like `hvd.init(process_sets="dynamic") --> In this case, process set can be altered dynamically

File: ignite/distributed/comp_models/horovod.py [hunk 9]
```diff
@@ -69,8 +68,10 @@ def __init__(self, backend: Optional[str] = None, **kwargs: Any) -> None:
 
         def _create_from_backend(self, backend: str, **kwargs: Any) -> None:
             self._backend = backend  # type: str
+            # comm = kwargs.get("comm", None)
```
Comment (vfdev-5) [inline ignite/distributed/comp_models/horovod.py:71]: Please revert this change as well

**Review by vfdev-5 — APPROVED** [review]
Looks ok, thanks @puhuk
